android/greengrass-android-component-base/component-library/src/main/java/com/aws/greengrass/android/component/core/GreengrassComponentService.java

```diff
@@ -0,0 +1,876 @@
+        RUNNING,        // stable
+        STOPPING,
+    }
+
+    /**
+     * Ensure component is stopped and update state to starting.
+     *
+     * @return true when component state was 'stopped'
+     */
+    boolean ensureStopped() {
+        synchronized (state) {
+            if (state.compareAndSet(ComponentState.STOPPED, ComponentState.STARTING)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Start thread, set 'started' state and wait until thread runs.
+     */
+    void doStart() {
+        // start thread and update state under lock
+        synchronized (state) {
+            workThread.get().start();
+            state.set(ComponentState.STARTED);
+        }
+
+        // waiting until thread is started
+        try {
+            threadStarted.await();
+        } catch (InterruptedException ignored) {
+        }
+    }
+
+    /**
+     * Update state to 'running' and notify other threads about stable state.
+     */
+    void setRunningState() {
+        // unblock start()
+        threadStarted.countDown();
+
+        synchronized (state) {
+            state.set(ComponentState.RUNNING);
+            state.notifyAll();
+        }
+    }
+
+    /**
+     * Checks got component arguments and environment.
+     * @return true if component already got arguments and environment
+     */
+    boolean hasEnvironment() {
+        synchronized (state) {
+            ComponentState localState = state.get();
+            return localState == ComponentState.STARTING
+                    || localState == ComponentState.STARTED
+                    || localState == ComponentState.RUNNING;
+        }
+    }
+
+    /**
+     * Checks is component stoppable.
+     * In case of transitive states STARTING or STARTED will waiting for max 300 ms until reach
+     *     any stable state.
+     * @return true when component can be stopped
+     */
+    boolean ensureStoppable() {
+        try {
+            // try 3 times to leave STARTING or STARTED
+            for (int i = 0; i < 3; i++) {
+                synchronized (state) {
+                    // can stop from STARTING, STARTED, RUNNING
+                    // bypass on STOPPED and STOPPING
+                    switch (state.get()) {
+                        case STOPPING:
+                        case STOPPED:
+                            return false;
+
+                        case RUNNING:
+                            state.set(ComponentState.STOPPING);
+                            state.notifyAll();
+                            return true;
+
+                        case STARTING:
+                        case STARTED:
+                        default:
+                            // waiting until state reach STOPPING, STOPPED or RUNNING
+                            state.wait(100);
+                            break;
+                    }
+                }
+            }
+        } catch (InterruptedException ignored) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)